### DISCUSSION TITLE
No version of chokidar, localhost가 연결을 거부

### DISCUSSION THREAD

**Discussion by kh-min7** [devrel-kr/HackaLearn#50] [2 Q&A - 클라우드 스킬 챌린지] [answered]
첫 npm start 입력 시에는 정상적으로 출력이 되었는데
그 뒤부터는

localhost이(가) 연결을 거부했습니다.
다음을 시도해 보세요.

란 창이 뜹니다.

어떻게 해결할 수 있을까요?

**Comment by minhoryang** [comment, marked as answer]
@kh-min7 해결을 위해서는 더 자세한 설명이 필요합니다. 예를들면
> 어떤 환경에서 어떤 코드를 작업하시는데, 어떤 명령 이후에 어떤 동작에 의해 어떤 화면이 어디서 뜨게된 건가요?

**Reply by kh-min7** [reply]
Azure Static Web Apps를 통한 사용자 인증 챕터중 연습 - 시작하기 단원

윈도우 vs code 에서 npm start 입력시 



You can now view react-app in the browser.

  Local:            http://localhost:3000

Note that the development build is not optimized.
To create a production build, use npm run build.

C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\watchpack\lib\chokidar.js:17
throw new Error(
^

Error: No version of chokidar is available. Tried chokidar@2 and chokidar@3.    
You could try to manually install any chokidar version.
chokidar@3: Error: Cannot find module 'chokidar'
Require stack:
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\watchpack\lib\chokidar.js
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\watchpack\lib\DirectoryWatcher.js
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\watchpack\lib\watcherManager.js
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\watchpack\lib\watchpack.js
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\webpack\lib\node\NodeWatchFileSystem.js
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\webpack\lib\node\NodeEnvironmentPlugin.js
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\webpack\lib\webpack.js
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\react-scripts\scripts\start.js
chokidar@2: Error: Cannot find module 'watchpack-chokidar2'
Require stack:
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\watchpack\lib\chokidar.js
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\watchpack\lib\DirectoryWatcher.js
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\watchpack\lib\watcherManager.js
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\watchpack\lib\watchpack.js
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\webpack\lib\node\NodeWatchFileSystem.js
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\webpack\lib\node\NodeEnvironmentPlugin.js
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\webpack\lib\webpack.js
- C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\react-scripts\scripts\start.js

\node_modules\watchpack\lib\chokidar.js:17:7)
    at Module._compile (internal/modules/cjs/loader.js:1072:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)   
    at Module.load (internal/modules/cjs/loader.js:937:32)
    at Function.Module._load (internal/modules/cjs/loader.js:778:12)
    at Module.require (internal/modules/cjs/loader.js:961:19)
    at require (internal/modules/cjs/helpers.js:92:18)
    at Object.<anonymous> (C:\Users\LG\Desktop\my-static-web-app-authn\react-app\node_modules\watchpack\lib\DirectoryWatcher.js:9:16)
Compiled successfully!

란 말이 뜨며 http://localhost:3000/ 가 실행되고 localhost가 연결을 거부했다고 인터넷 창이 뜹니다.

**Reply by kooljay82** [reply]
npm version이 어떻게 되시나요??

**Reply by minhoryang** [reply]
오! 이 traceback에서 중요한 문구는 `Error: No version of chokidar is available. Tried chokidar@2 and chokidar@3.`로 보입니다.
- `watchpack`이 `chokidar`를 사용하려 하는데.
- `chokidar` 라이브러리가 없다는 문제네요.

(참고: [watchpack npm](https://www.npmjs.com/package/watchpack), [chokidar npm](https://www.npmjs.com/package/chokidar))

- (추측) 우선 현재 환경이 Windows 라서 발생하는 문제로 예상됩니다. package-lock.json을 생성하는 시점에는 Windows가 아니었는데, package-lock.json을 사용해 패키지를 설치하려하니, Windows환경에서만(!!) 필요한 라이브러리들이 설치가 안된 것으로 보입니다.

제 제안은
- 우선 package-lock.json을 지우고 다시 `npm install`하여, `npm start`가 잘 되는지 확인한다.
- 위의 방향으로 문제가 해결된다면 ok, 안된다면 `npm install --save-dev chokidar`하여 chokidar를 포함시킨다.
- 빌드가 잘 되는지 확인 후, 내 컴퓨터에서 잘된다면, 커밋 후 푸시해본다.

입니다.

**Reply by minhoryang** [reply]
https://github.com/MicrosoftDocs/mslearn-staticwebapp/blob/main/react-app/package-lock.json 를 확인해보니 `chokidar`가 존재하는것을 확인할 수 있었습니다. package-lock.json에서 적시되어있다면, 플랫폼 문제(OS/Node version)가 아니라면 설치가 될거라고 예상할 수 있습니다.

- 현재 무슨 버전의 node를 쓰는지 먼저 확인해주셔야 할 것 같습니다. `chokidar@2`는 node 8이상, `chokidar@3`은 node 15이상에서 작동한다고 합니다. (package-lock.json을 통한 추측)
- 위의 메세지를 통해 유추해보면 `npm install`시에도 몇가지 문제가 있었을 것 같습니다. `node_modules`를 지운 후 `npm install`을 실행하여, 그 창에서 발생한 메세지를 파일로 저장하여 여기에 첨부해주시면 분석에 도움이 될 것 같습니다.

**Reply by minhoryang** [reply]
중요한건 현재 과제 가이드라인에는 Node LTS버전을 써달라고 되어있으며, 현재 LTS버전은 14입니다. Node 14버전을 다운받아 설치하셔야합니다. https://nodejs.org/en/download/

**Reply by kh-min7** [reply]
Node 버전 확인해보니 14여서 말씀해주신대로 순서대로 해봤습니다.
npm install --save-dev chokidar 를 실행하고 npm start를 하니 정상적으로 원하는 결과가 나왔네요. 

정말 감사합니다!

**Reply by kh-min7** [reply]
> npm version이 어떻게 되시나요??

npm은 7.20.3
node는 14.17.4 였었네요!

**Reply by minhoryang** [reply]
@kh-min7 오 그렇다면, package-lock.json 으로 설치가 잘 되었어야했는데, 안된게 조금 신기하네요. 고생하셨습니다.
혹시 나중에 Azure Static Web App의 빌드과정에서도 별 문제가 없는지 한번 확인해주시면 감사하겠습니다.

**Comment by minhoryang** [comment]
이 질문의 제목을 `[react-app] npm start -> No version of chokidar, localhost가 연결을 거부`라고 수정한다면, 다른분들이 더 빠르게 도움을 받을 수 있을 것 같습니다.

**Reply by kh-min7** [reply]
넵 알겠습니다!
